lib/ClangImporter/ClangDerivedConformances.cpp

```diff
@@ -1364,3 +1364,217 @@ void swift::conformToCxxSpanIfNeeded(ClangImporter::Implementation &impl,
+  clang::Expr *synthCtorExpr = synthCtorExprResult.get();
+
+  SmallVector<clang::Expr *, 2> args = {lhsParamRefExpr, rhsParamRefExpr};
+  auto call =
+      clangSema.BuildCallExpr(nullptr, synthCtorExpr, clang::SourceLocation(),
+                              args, clang::SourceLocation());
+
+  if (!call.isUsable())
+    return false;
+
+  auto opBody = clang::ReturnStmt::Create(clangCtx, clang::SourceLocation(),
+                                          call.get(), nullptr);
+  opDecl->setBody(opBody);
+
+  impl.synthesizedAndAlwaysVisibleDecls.insert(opDecl);
+  auto lookupTable1 = impl.findLookupTable(classDecl);
+  addEntryToLookupTable(*lookupTable1, opDecl, impl.getNameImporter());
+  auto owningModule = impl.getClangOwningModule(classDecl);
+  auto lookupTable2 = impl.findLookupTable(owningModule);
+  if (lookupTable1 != lookupTable2)
+    addEntryToLookupTable(*lookupTable2, opDecl, impl.getNameImporter());
+  return true;
+}
+
+static clang::ClassTemplateSpecializationDecl *
+lookupAndSpecializeFunctionObject(ClangImporter::Implementation &impl,
+                                  const clang::CXXRecordDecl *clangDecl,
+                                  std::string name) {
+  clang::ASTContext &Ctx = impl.getClangASTContext();
+  clang::Sema &clangSema = impl.getClangSema();
+
+  clang::NamespaceDecl *stdNS = clangSema.getStdNamespace();
+  // maybe the C++ header didn't include standard library?
+  if (!stdNS)
+    return nullptr;
+
+  clang::DeclarationName declName =
+      Ctx.DeclarationNames.getIdentifier(&Ctx.Idents.get(name));
+
+  clang::ClassTemplateDecl *tmplDecl =
+      stdNS->lookup(declName).find_first<clang::ClassTemplateDecl>();
+
+  if (!tmplDecl)
+    return nullptr;
+
+  // find specialization to the current clangDecl
+  void *insertPos = nullptr;
+  return tmplDecl->findSpecialization(
+      {clang::TemplateArgument(
+          clangDecl->getASTContext().getTypeDeclType(clangDecl))},
+      insertPos);
+}
+
+void swift::conformToHashableIfNeeded(ClangImporter::Implementation &impl,
+                                      SwiftDeclSynthesizer &synthesizer,
+                                      NominalTypeDecl *decl,
+                                      const clang::CXXRecordDecl *clangDecl) {
+  PrettyStackTraceDecl trace("conforming to Hashable", decl);
+
+  ASSERT(decl);
+  ASSERT(clangDecl);
+
+  auto *stdHashSpec =
+      lookupAndSpecializeFunctionObject(impl, clangDecl, "hash");
+  auto *stdETSpec =
+      lookupAndSpecializeFunctionObject(impl, clangDecl, "equal_to");
+
+  // we bail if [std::hash<>] couldn't be found
+  if (!stdHashSpec)
+    return;
+
+  if (!stdETSpec && !getEqualEqualOperator(decl))
+    return;
+
+  auto *stdHash = impl.importDecl(stdHashSpec, impl.CurrentVersion);
+  if (!stdHash)
+    return;
+
+  if (auto *stdHashStruct = dyn_cast<StructDecl>(stdHash)) {
+    FuncDecl *hashFunc = synthesizer.makeHashFunc(decl, stdHashStruct);
+    decl->addMember(hashFunc);
```

Comment:
I'm wondering whether we should wait to import `std::hash`, synthesize the hash function, and add this member to `decl` if and only if we synthesize the conformance? Or is it intentional that we add a hash method even if we can't find a suitable `==(_:_:)`?